blog/2018/12/22/use-carton-for-your-mojolicious-app-deployment/index.markdown

```diff
@@ -0,0 +1,155 @@
+
+Cpanm is able to install modules in a specific folder (thank you, [local::lib](https://metacpan.org/pod/local::lib)), but wouldn’t it be convenient to install our dependencies in the directory of our application?
+We would always know where our dependencies are.
+
+## Here comes Carton
+
+[Carton](https://metacpan.org/pod/Carton) is Perl module dependency manager. Think `bundler` in Ruby. Think `npm` in Node.js.
+
+Like `npm` does, Carton install the dependencies in the directory of the application.
+
+### Deployment
+
+First, install Carton:
+
+    cpanm Carton
+
+Then, we can install our dependencies with:
+
+    carton install
+
+Our dependencies will be installed in a directory named `local`.
+But there is more: Carton will generate a `cpanfile.snapshot` file, containing the exact version of our dependencies, allowing us to enforce those exact version (ship it with your application).
+
+In our `cpanfile` example, we asked for a Mojolicious version superior or equal to 7.0 and inferior to 8.0.
+Between the installation on our development server and the installation on the production server, some versions may have been published.
+Let’s say that we have Mojolicious 7.77 in our development environment and 7.90 and that something has changed, which leads to problems (for example, the delay helper from [Mojolicious::Plugin::DefaultHelpers](https://mojolicious.org/perldoc/Mojolicious/Plugin/DefaultHelpers) has been [deprecated in 7.78](https://github.com/mojolicious/mojo/blob/47d1369fd11b09af47a76f7f7192985a30ce2409/Changes#L243) and [removed in 7.90](https://github.com/mojolicious/mojo/blob/47d1369fd11b09af47a76f7f7192985a30ce2409/Changes#L150).
+
+Both 7.77 and 7.90 versions are in our range, but our application does not work on the production server… we need to make the production environment as identical as possible as the development one.
+
+For that, since we have a `cpanfile.snapshot` file from our development server, we can do:
+
+    carton install --deployment
+
+### Features
+
+Per default, `carton install` will install all the *features* dependencies, but we can deactivate some:
+
+    carton install --deployment --without-feature mysql
+
+In order to provide the correct version for all modules, even the optional ones, do a `carton install` on the development server, and use `--without-feature` only while deploying your application: you need to have a `cpanfile.snapshot` containing all modules.
+
+### Start your application
+
+In order to be able to use the `local` directory containing the dependencies, you can prefix your commands with `carton exec`.
+So, to start a Mojolicious application with the built-in server [hypnotoad](https://mojolicious.org/perldoc/Mojo/Server/Hypnotoad), do:
+
+    carton exec hypnotoad script/my_application
+
+That works for all that you can do with your application. Example:
+
+    carton exec script/my_application routes
```

Comment:
I find I often have to use `carton exec -- ` so that arguments passed to the script don't get taken by carton. Maybe mention that? Up to you.

Reply:
Yes, sure! :+1: